src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```
@@ -200,60 +201,60 @@ boolean isOpen() {
 			}
 			else {
-				return CompletableFutureUtil.timedRunAsync(timeout, () -> {
-					RemoteEndpoint.Basic remote = this.session.getBasicRemote();
-					
-					try {
-						if (mode.equals(WriteMode.TEXT)) {
-							String text = data.toString();
-							RelayLogger.logEvent("writingBytes", this, "text");
-							Writer writer = remote.getSendWriter();
-							writer.write(text);
-							writer.close();
-							RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(text.length()));
-						}
-						else {
-							ByteBuffer bytes = null;
-							
-							RelayLogger.logEvent("writingBytes", this, "binary");
-							if (data instanceof byte[]) {
-								bytes = ByteBuffer.wrap((byte[]) data);
-							} 
-							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
+				RemoteEndpoint.Basic remote = this.session.getBasicRemote();
+				RelayLogger.logEvent("writingBytes", this, mode.toString());
+				
+				// The websocket API will throw if multiple sends are attempted on the same websocket simultaneously
+				// Therefore a lock is used on the send operation
+				return this.socketLock.acquireThenCompose(timeout, () -> {
+					return CompletableFutureUtil.timedRunAsync(timeout, () -> {
+						try {
+							if (mode.equals(WriteMode.TEXT)) {
+								String text = data.toString();
+								remote.sendText(text, isEnd);
+								RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(text.length()));
 							}
-							else if (data instanceof String){
-								bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+							else {
+								ByteBuffer bytes = null;
+								if (data instanceof byte[]) {
+									bytes = ByteBuffer.wrap((byte[]) data);
+								} 
+								else if (data instanceof ByteBuffer) {
+									bytes = (ByteBuffer) data;
+								}
+								else if (data instanceof String){
+									bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+								}
```

Isn't it possible for an Object to not fall into any of these if/else-if combinations? Ideally this API should throw instead of doing an empty send in that case.